dtls.c

```diff
@@ -2016,8 +2081,8 @@ static int
 dtls_send_certificate_ecdsa(dtls_context_t *ctx, dtls_peer_t *peer,
 			    const dtls_ecdsa_key_t *key)
 {
-  uint8 buf[DTLS_CE_LENGTH];
   uint8 *p;
+  uint8 buf[DTLS_CE_LENGTH];
```

Comment:
I do not see why this needs to be re-ordered.

Reply:
Original code was of the form
````
  uint8 buf[some_size];
  uint8 *p;
  /* Start message construction at beginning of buffer. */
  p = buf;
  ... multiple instances of the next 2 lines
  memcpy(p, some_data, size)
  p += size;
  ...
  assert(p <= (buf + sizeof(buf)));
````
If the size of `buf `is too small, with the original ordering, after several `memcpy()`, `p` will point to (almost) the end of `buf `and the next `memcpy()` will actually overwrite `p` as it immediately follows `buf`.  I ran into this issue as `p` ended up being a large value and so the `assert()` did not trigger and other memory was corrupted.
Defining `p` before `buf` in the stack meant I caught my `buf` sizing error.

Reply:
Hell!
if you reorder that, I hope you don't kill your call-stack! Otherwise a return may end in "no-mans-land".

Reply:
Well, the idea is to fix potential buffer overflows eventually.

Reply:
Hmm, which ever way round these are defined, an overflow of `buf` will cause issues.

I have just disassembled the x86 code with them being defined in either order - in both cases `p` is higher in the stack than `buf ` for the couple of functions I checked - so compiler is doing some re-arranging here (likely to be down to 8 byte alignment and making use of non-aligned space).  I will revert out these `p `/ `buf `re-ordering changes as they appear to be making no difference and we are pretty confident that the `buf `sizes are OK.

Reply:
Yes, it is on the optimizer to decide. In theory, `p` could even end up in a register. Anyway, if you compiled with `-O0`, you should see the order you have specified in the code.

Reply:
Decades ago I used macro's and a small set of functions to but a "mark" around buffers and verify that. That fails to detect "jumps in buffer" (e.g. p *= 128; *p= 0). But helps to locate a couple of violations. It's too long ago, that I was really "up-to-date" in C, so maybe, in the meantime, compilers offers such things.

But tricks like that "order of buffer and point", didn't work in my experience of the past.

Reply:
It is likely I was working with no optimization as I was debugging the code when I stumbled into this issue.